### PR TITLE
Use a vault to store public certs

### PR DESCRIPTION
**Pull request by asalkeld**
```release-note
plugin now reads apiserver and router certificates from keyvault.  API changes:
1. Properties.ServicePrincipalProfile removed in favour of Properties.MasterServicePrincipalProfile and Properties.WorkerServicePrincipalProfile.
2. Properties.APICertProfile and Properties.RouterProfiles.CertProfile added, must contain links to applicable keyvault secrets.
3. ContextKeyVaultClientAuthorizer context key added: must be set to an Authorizer which can read secrets from the customer keyvault (see azureclient.NewAuthorizer)
```

cc @jim-minter

### CODE REVIEW

**Comment by asalkeld** [comment]
relates to #1125

File: pkg/util/azureclient/graphrbac_addons.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import (
+	"context"
+	"fmt"
+)
+
+type ServicePrincipalsClientAddons interface {
+	ObjectIDforApplicationID(ctx context.Context, appID string) (string, error)
+}
+
+func (spc *servicePrincipalsClient) ObjectIDforApplicationID(ctx context.Context, appID string) (string, error) {
```
Comment (jim-minter) [inline pkg/util/azureclient/graphrbac_addons.go]: 1) reluctant to use addons for this - please can you put in a library function outside of pkg/util/azureclient
2) turns out this is also implemented in UpdateAADAppSecret() - please can you deduplicate?

File: pkg/util/aadapp/aadapp.go [hunk 2]
```diff
@@ -25,7 +23,27 @@ func UpdateAADAppSecret(ctx context.Context, appClient azureclient.RBACApplicati
 	if len(apps) != 1 {
 		return "", fmt.Errorf("found %d applications, should be 1", len(apps))
 	}
-	app := apps[0]
+	return *apps[0].ObjectID, nil
+}
+
+// GetObjectIDUsingSPClient find the ObjectID for the application
+func GetObjectIDUsingSPClient(ctx context.Context, spc azureclient.ServicePrincipalsClient, appID string) (string, error) {
```
Comment (asalkeld) [inline pkg/util/aadapp/aadapp.go]: @jim-minter I tried using either of these two functions for both use cases and kept getting 403's. It was driving me crazy maybe someone can figure it out? at least they are together so we know about them...

Comment (jim-minter) [reply pkg/util/aadapp/aadapp.go]: 2 different objects - AAD applications (e.g. `az ad app list`) and service principals (e.g. `az ad sp list`).  Each have an object ID; each also have an appID field.

Comment (jim-minter) [reply pkg/util/aadapp/aadapp.go]: So my advice that the same function could be used for both was wrong - sorry.

**Comment by jim-minter** [comment]
/hold
This PR is doing a great deal of unnatural things.  I'm working on a branch to deal with them [here](https://github.com/jim-minter/openshift-azure/commit/new-vault-stuff).

File: cmd/startup/startup.go [hunk 3]
```diff
@@ -68,15 +57,11 @@ func (s *startup) writeTemplatedFiles() error {
...
-				"PrivateKeyFromVault": s.privateKeyFromVault,
-			}, s.cs, map[string]interface{}{
-				"Hostname":    hostname,
-				"DNSHostname": domainname,
-				"TestConfig":  tc,
-			})
+		b, err := template.Template(string(templateFile), nil, s.cs, map[string]interface{}{
+			"Hostname":    hostname,
+			"DNSHostname": domainname,
+			"TestConfig":  tc,
```
Comment (jim-minter) [inline cmd/startup/startup.go]: @charlesakalugwu @asalkeld this should come out when #1175 merges

File: pkg/api/plugin.go [hunk 4]
```diff
@@ -11,8 +11,9 @@ import (
 type ContextKey string
 
 const (
-	ContextKeyClientAuthorizer ContextKey = "ClientAuthorizer"
-	ContextAcceptLanguages     ContextKey = "AcceptLanguages"
+	ContextKeyClientAuthorizer      ContextKey = "ClientAuthorizer"
+	ContextKeyVaultClientAuthorizer ContextKey = "VaultClientAuthorizer"
```
Comment (jim-minter) [inline pkg/api/plugin.go]: @asalkeld need to highlight this new key and how to use it to MSFT in the release notes

File: pkg/fakerp/admin_handlers.go [hunk 5]
```diff
@@ -138,7 +138,7 @@ func (s *Server) handleRestore(w http.ResponseWriter, req *http.Request) {
 	etcdContainer := bsc.GetContainerReference(cluster.EtcdBackupContainerName)
 
 	blob := etcdContainer.GetBlobReference(blobName)
-	exists, err := blob.Exists()
+	exists, err := blob.Exists() // TODO: this check should be within the plugin
```
Comment (jim-minter) [inline pkg/fakerp/admin_handlers.go]: @asalkeld please can you open a follow-up issue for this, tagged GA?

File: pkg/fakerp/customer_handlers.go [hunk 6]
```diff
@@ -43,6 +43,13 @@ func (s *Server) handleDelete(w http.ResponseWriter, req *http.Request) {
 		return
 	}
 
+	// TODO: we shoult not be creating clients left right and center - all in one place
```
Comment (jim-minter) [inline pkg/fakerp/customer_handlers.go]: @asalkeld please can you open a follow-up issue for this?

File: pkg/fakerp/vault.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+				DNSNames: []string{
+					cs.Properties.RouterProfiles[0].PublicSubdomain,
+					"*." + cs.Properties.RouterProfiles[0].PublicSubdomain,
+				},
+				ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+			},
+		},
+		// Do not attempt to make the OpenShift console certificate self-signed
+		// if cs.Properties == cs.FQDN:
+		// https://github.com/openshift/openshift-azure/issues/307
```
Comment (jim-minter) [inline pkg/fakerp/vault.go]: This issue risks blocking integration.  Need to discuss with MSFT.

**Comment by jim-minter** [comment]
@asalkeld have pushed to your branch - ptal.
am marking lgtm (let's see if tests pass) but there are follow-ups.
if you're happy, do a hold cancel

/lgtm

**Comment by jim-minter** [comment]
> @asalkeld have pushed to your branch - ptal.
@asalkeld the commit comment should help see what's going on.

**Pull request merged**